File: hammerspoon/cr/matcher.lua
-- cr.matcher — referent binding + presence predicate.
--
-- v1 is deictic-only and deliberately LLM-free: a referent is "whatever was on
-- screen when the reminder was created", and presence is a deterministic match
-- against later snapshots. Fully replayable; the LLM earns its way in later,
-- only where determinism runs out (non-deictic conditions, ambiguity).

local M = {}

-- Strip spinner/animation glyphs from window titles before comparing.
-- Dogfooding finding #1: kitty renders Claude Code's Braille spinner in the
-- title bar (⠐ → ⠂ → …), which made every frame look like a context change.
local SPINNER = {
  [0x2733] = true, [0x2731] = true, [0x273B] = true, [0x273D] = true, -- ✳ ✱ ✻ ✽
  [0x25CF] = true, [0x23FA] = true, [0x25CB] = true,                  -- ● ⏺ ○
}

function M.normalizeTitle(s)
  if type(s) ~= "string" or s == "" then return nil end
  local ok, out = pcall(function()
    local parts = {}
    for _, cp in utf8.codes(s) do
      local isBraille = cp >= 0x2800 and cp <= 0x28FF
      if not isBraille and not SPINNER[cp] then
        parts[#parts + 1] = utf8.char(cp)
      end
    end
    return table.concat(parts)
  end)
  if not ok then out = s end -- invalid utf8: compare raw rather than crash
  out = out:gsub("%s+", " "):gsub("^%s+", ""):gsub("%s+$", "")
  if out == "" then return nil end
  return out
end

-- Canonical key for a URL. YouTube gets a special case — the video id is the
-- identity; timestamps and playlist params are noise. Everything else keeps
-- its query string (a Google search IS its query) but drops the fragment.
function M.urlKey(url)
  if type(url) ~= "string" or url == "" then return nil end
  local host = url:match("^https?://([^/]+)") or ""
  if host:find("youtube%.com") then
    local vid = url:match("[?&]v=([%w_%-]+)")
    if vid then return "yt:" .. vid end
  end
  local short = url:match("^https?://youtu%.be/([%w_%-]+)")
  if short then return "yt:" .. short end
  return (url:gsub("#.*$", ""):gsub("/$", ""))
end

-- Bind a referent from a snapshot: "this" = what's on screen right now.
function M.bind(snap)
  if not snap then return nil end
  local title = M.normalizeTitle(snap.tab or snap.title)
  local urlKey = M.urlKey(snap.url)
  if not title and not urlKey then return nil end
  return {
    bundle    = snap.bundle,
    app       = snap.app,
    urlKey    = urlKey,
    titleCore = title,
    boundAt   = os.time(),
    label     = (snap.app or "?") .. " — " .. (title or snap.url or "untitled"),
  }
end

local function titleMatch(core, s)
  if not core or not s then return false end
  local n = M.normalizeTitle(s)
  if not n then return false end
  local a, b = n:lower(), core:lower()
  return a == b or a:find(b, 1, true) ~= nil or b:find(a, 1, true) ~= nil
end

-- Presence: is the referent visible in this snapshot?
function M.matches(ref, snap)
  if not ref or not snap then return false end
  if ref.urlKey then
    -- URL-bound (browser): the active tab is the identity
    if M.urlKey(snap.url) == ref.urlKey then return true end
  elseif ref.bundle and snap.bundle == ref.bundle then
    -- title-bound (everything else): same app + matching normalized title
    if titleMatch(ref.titleCore, snap.tab or snap.title) then return true end
  end
  -- media fallback: a bound video still playing in a background tab is not
  -- "done" — requires media-control to be installed, degrades to false without
  if ref.titleCore and snap.media and snap.media.playing
    and titleMatch(ref.titleCore, snap.media.title) then
    return true
  end
  return false
end

return M

File: windows/crw/matcher.py
"""Referent binding + presence predicate — port of cr/matcher.lua.

Deictic-only and deliberately LLM-free: a referent is "whatever was on screen
when the reminder was created", and presence is a deterministic match against
later snapshots. On Windows the app identity is the process executable
("Code.exe") where macOS used the bundle id; browser tabs are title-bound for
now (Chromium puts the tab title in the window title), with url_key kept for
when a URL source lands.
"""

from __future__ import annotations

import re

# Spinner/animation glyphs some apps render into their titles — every frame
# looked like a context change until these were stripped (dogfooding finding
# #1 on the Mac; terminals do the same on Windows).
_SPINNER = set("✳✱✻✽●⏺○")


def normalize_title(s):
    if not isinstance(s, str) or not s:
        return None
    out = "".join(
        ch for ch in s
        if not (0x2800 <= ord(ch) <= 0x28FF) and ch not in _SPINNER)
    out = re.sub(r"\s+", " ", out).strip()
    return out or None


def url_key(url):
    """Canonical key for a URL. YouTube gets a special case — the video id is
    the identity. Everything else keeps its query (a Google search IS its
    query) but drops the fragment."""
    if not isinstance(url, str) or not url:
        return None
    m = re.match(r"^https?://([^/]+)", url)
    host = m.group(1) if m else ""
    if "youtube.com" in host:
        v = re.search(r"[?&]v=([\w-]+)", url)
        if v:
            return "yt:" + v.group(1)
    short = re.match(r"^https?://youtu\.be/([\w-]+)", url)
    if short:
        return "yt:" + short.group(1)
    return re.sub(r"#.*$", "", url).rstrip("/")


def bind(snap):
    """Bind a referent from a snapshot: "this" = what's on screen right now."""
    if not snap:
        return None
    title = normalize_title(snap.get("tab") or snap.get("title"))
    ukey = url_key(snap.get("url"))
    if not title and not ukey:
        return None
    return {
        "exe": snap.get("exe"),
        "app": snap.get("app"),
        "urlKey": ukey,
        "titleCore": title,
        "boundAt": None,  # stamped by the caller (reminders.add)
        "label": f"{snap.get('app') or '?'} — {title or snap.get('url') or 'untitled'}",
    }


def _title_match(core, s):
    if not core or not s:
        return False
    n = normalize_title(s)
    if not n:
        return False
    a, b = n.lower(), core.lower()
    return a == b or b in a or a in b


def matches(ref, snap) -> bool:
    """Presence: is the referent visible in this snapshot?"""
    if not ref or not snap:
        return False
    if ref.get("urlKey"):
        if url_key(snap.get("url")) == ref["urlKey"]:
            return True
    elif ref.get("exe") and snap.get("exe") == ref["exe"]:
        if _title_match(ref.get("titleCore"), snap.get("tab") or snap.get("title")):
            return True
    # media fallback: a bound video still playing in the background is not
    # "done" — needs a media source (winsdk); degrades to False without one
    media = snap.get("media")
    if (ref.get("titleCore") and media and media.get("playing")
            and _title_match(ref["titleCore"], media.get("title"))):
        return True
    return False
